### ISSUE TITLE
Question on the use of OHEMSampler

### ISSUE THREAD

**Issue by caiqi** [dmlc/gluon-cv#225]
If the negative_mining_ratio in SSDDefaultTrainTransform is set to 3, it will call OHEMSampler. However the cls_preds in SSDTargetGenerator is None since targets are pre-computed before cls_preds are computed. I tried to feed the cls_preds into SSDTargetGenerator during training. But it throws the error 

Check failed: (*rit)->inputs.size() == input_grads.size() (2 vs. 1) Gradient function not returning enough gradient

The code is like this:
            cls_targets_list = []
            box_targets_list = []
            with autograd.record():
                cls_preds = []
                box_preds = []
                for idx, x in enumerate(data):
                    cls_pred, box_pred, anchors = net(x)
                    cls_preds.append(cls_pred)
                    box_preds.append(box_pred)
                    cls_targets_, box_targets_, box_masks = target_generator(anchors,cls_pred,grd_box,grd_cls )
                    cls_targets_list.append(cls_targets_)
                    box_targets_list.append(box_targets_)

                sum_loss, cls_loss, box_loss = mbox_loss(
                    cls_preds, box_preds, cls_targets_list, box_targets_list)
                autograd.backward(sum_loss)

**net** is a wrapper of a symbol with pretrained model from mxnet-ssd.

It seems that some operation in SSDTargetGenerator are differential. Is there any way to use OHEMSampler for generating targets? Thanks! @zhreshold

**Comment by zhreshold** [comment]
You can wrap the target generator with 
```
with autograd.pause():
   xxx
```

Or use F.stop_gradient to block some backward paths.

**Comment by pribadihcr** [comment]
Hi @caiqi ,

how do you modify SSDDefaultTrainTransform, since in self._target_generato, the cls_preds is None

**Comment by caiqi** [comment]
@pribadihcr I modified the code in SSDDefaultTrainTransform to return gt_ids, gt_bboxes. The targets are generated after getting the prediction.
